File: tools/galaxy_ai_voice_subtitle_studio/app/server/routers/project_graph.py
from __future__ import annotations

from dataclasses import asdict
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from ...project_graph.models import AssetReference, HandoffRequest, NodeRequest
from ...project_graph.runtime import project_graph_service
from ...project_graph.service import ProjectGraphService, workspace_catalog


router = APIRouter(prefix="/api/project-graph", tags=["project-graph"])


class AssetBody(BaseModel):
    asset_id: str = Field(min_length=1)
    role: str = Field(min_length=1)
    path_hint: str = ""
    ownership: str = "linked"
    fingerprint: str = ""
    derived_from: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

    def contract(self) -> AssetReference:
        return AssetReference(
            asset_id=self.asset_id,
            role=self.role,
            path_hint=self.path_hint,
            ownership=self.ownership,
            fingerprint=self.fingerprint,
            derived_from=tuple(self.derived_from),
            metadata=self.metadata,
        )


class NodeBody(BaseModel):
    project_id: str = Field(min_length=1)
    workspace: str = Field(min_length=1)
    owner_id: str = Field(min_length=1)
    label: str = ""
    revision: int = Field(0, ge=0)
    assets: list[AssetBody] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

    def contract(self) -> NodeRequest:
        return NodeRequest(
            project_id=self.project_id,
            workspace=self.workspace,
            owner_id=self.owner_id,
            label=self.label,
            revision=self.revision,
            assets=tuple(item.contract() for item in self.assets),
            metadata=self.metadata,
        )


class HandoffBody(BaseModel):
    project_id: str = Field(min_length=1)
    source_node_id: str = Field(min_length=1)
    target_workspace: str = Field(min_length=1)
    input_asset_ids: list[str] = Field(default_factory=list)
    payload: dict[str, Any] = Field(default_factory=dict)
    handoff_id: str = ""


class ReturnBody(BaseModel):
    target_node: NodeBody | None = None
    target_node_id: str = ""
    output_asset_ids: list[str] = Field(default_factory=list)


def _service(request: Request) -> ProjectGraphService:
    configured = getattr(request.app.state, "settings_path", None)
    return project_graph_service(Path(configured) if configured is not None else None)


@router.get("/workspaces")
def list_workspaces() -> list[dict[str, Any]]:
    return [
        {
            "id": item.workspace_id,
            "label": item.label,
            "route": item.route,
            "targets": list(item.targets),
        }
        for item in workspace_catalog()
    ]


@router.get("/projects/{project_id}")
def get_project_graph(project_id: str, request: Request) -> dict[str, Any]:
    try:
        return asdict(_service(request).get_graph(project_id))
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error


@router.post("/nodes")
def upsert_node(body: NodeBody, request: Request) -> dict[str, Any]:
    try:
        return asdict(_service(request).upsert_node(body.contract()))
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error


@router.post("/handoffs")
def create_handoff(body: HandoffBody, request: Request) -> dict[str, Any]:
    try:
        return asdict(
            _service(request).create_handoff(
                HandoffRequest(
                    project_id=body.project_id,
                    source_node_id=body.source_node_id,
                    target_workspace=body.target_workspace,
                    input_asset_ids=tuple(body.input_asset_ids),
                    payload=body.payload,
                    handoff_id=body.handoff_id,
                )
            )
        )
    except KeyError as error:
        raise HTTPException(status_code=404, detail="Không tìm thấy node nguồn.") from error
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error


@router.get("/handoffs/{handoff_id}")
def get_handoff(handoff_id: str, request: Request) -> dict[str, Any]:
    try:
        return asdict(_service(request).get_handoff(handoff_id))
    except KeyError as error:
        raise HTTPException(status_code=404, detail="Không tìm thấy handoff.") from error


@router.post("/handoffs/{handoff_id}/open")
def open_handoff(handoff_id: str, request: Request) -> dict[str, Any]:
    try:
        return asdict(_service(request).open_handoff(handoff_id))
    except KeyError as error:
        raise HTTPException(status_code=404, detail="Không tìm thấy handoff.") from error
    except ValueError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error


@router.post("/handoffs/{handoff_id}/return")
def return_handoff(handoff_id: str, body: ReturnBody, request: Request) -> dict[str, Any]:
    try:
        return asdict(
            _service(request).return_handoff(
                handoff_id,
                target_node=body.target_node.contract() if body.target_node else None,
                target_node_id=body.target_node_id,
                output_asset_ids=tuple(body.output_asset_ids),
            )
        )
    except KeyError as error:
        raise HTTPException(status_code=404, detail="Không tìm thấy handoff.") from error
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error

File: tools/galaxy_ai_voice_subtitle_studio/frontend/src/api/projectGraph.ts
import { apiJson } from './client'

export type AssetOwnership = 'managed' | 'linked' | 'generated'
export type HandoffStatus = 'pending' | 'opened' | 'returned'

export interface ProjectAssetReference {
  asset_id: string
  role: string
  path_hint: string
  ownership: AssetOwnership
  fingerprint: string
  derived_from: string[]
  metadata: Record<string, unknown>
}

export interface ProjectGraphNode {
  node_id: string
  project_id: string
  workspace: string
  owner_id: string
  label: string
  route: string
  revision: number
  assets: ProjectAssetReference[]
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ProjectHandoff {
  handoff_id: string
  project_id: string
  source_node_id: string
  source_workspace: string
  source_revision: number
  source_route: string
  target_workspace: string
  target_route: string
  target_node_id: string
  status: HandoffStatus
  input_asset_ids: string[]
  output_asset_ids: string[]
  payload: Record<string, unknown>
  created_at: string
  opened_at: string
  returned_at: string
}

export interface ProjectGraph {
  project_id: string
  nodes: ProjectGraphNode[]
  handoffs: ProjectHandoff[]
  updated_at: string
}

export interface ProjectWorkspaceSpec {
  id: string
  label: string
  route: string
  targets: string[]
}

export interface ProjectNodeInput {
  project_id: string
  workspace: string
  owner_id: string
  label: string
  revision?: number
  assets?: ProjectAssetReference[]
  metadata?: Record<string, unknown>
}

export const fetchProjectGraph = (projectId: string) =>
  apiJson<ProjectGraph>(`/api/project-graph/projects/${encodeURIComponent(projectId)}`)

export const fetchProjectWorkspaces = () =>
  apiJson<ProjectWorkspaceSpec[]>('/api/project-graph/workspaces')

export const upsertProjectNode = (body: ProjectNodeInput) =>
  apiJson<ProjectGraphNode>('/api/project-graph/nodes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })

export const createProjectHandoff = (body: {
  project_id: string
  source_node_id: string
  target_workspace: string
  input_asset_ids?: string[]
  payload?: Record<string, unknown>
}) => apiJson<ProjectHandoff>('/api/project-graph/handoffs', {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(body),
})

export const openProjectHandoff = (handoffId: string) =>
  apiJson<ProjectHandoff>(`/api/project-graph/handoffs/${encodeURIComponent(handoffId)}/open`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: '{}',
  })

export const returnProjectHandoff = (
  handoffId: string,
  body: { target_node_id?: string; output_asset_ids: string[] },
) => apiJson<ProjectHandoff>(
  `/api/project-graph/handoffs/${encodeURIComponent(handoffId)}/return`,
  {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  },
)
